test/plugin/base_test.rb

```diff
@@ -305,20 +305,83 @@ def test_structured_payload_malformatted_log
   def test_structured_payload_json_log
     setup_gce_metadata_stubs
     setup_logging_stubs do
-      d = create_driver
-      json_string = '{"msg": "test log entry 0", "tag2": "test", "data": 5000}'
+      d = create_driver(APPLICATION_DEFAULT_CONFIG)
+      json_string = '{"msg": "test log entry 0", "tag2": "test", ' \
+                    '"data": 5000, "some_null_field": null}'
+      d.emit('message' => 'notJSON ' + json_string)
+      d.emit('log' => 'notJSON ' + json_string)
+      d.emit('msg' => 'notJSON ' + json_string)
+      d.emit('message' => json_string)
+      d.emit('log' => json_string)
+      d.emit('msg' => json_string)
+      d.emit('message' => "  \r\n \t" + json_string)
+      d.emit('log' => "  \r\n \t" + json_string)
+      d.emit('msg' => "  \r\n \t" + json_string)
+      d.run
+    end
+    log_index = 0
+    verify_log_entries(9, COMPUTE_PARAMS, '') do |entry|
+      log_index += 1
+      if log_index % 3 == 1
+        assert entry.key?('textPayload'),
+               "Entry ##{log_index} #{entry} did not have textPayload"
+      else
+        assert entry.key?('jsonPayload'),
+               "Entry ##{log_index} #{entry} did not have jsonPayload"
+        fields = get_fields(entry['jsonPayload'])
+        assert !fields.key?('tag2'), 'Did not expect tag2'
+        assert !fields.key?('data'), 'Did not expect data'
+        assert !fields.key?('some_null_field'),
+               'Did not expect some_null_field'
+      end
+    end
+  end
+
+  def test_structured_payload_json_log_detect_json
+    setup_gce_metadata_stubs
+    setup_logging_stubs do
+      d = create_driver(DETECT_JSON_CONFIG)
+      json_string = '{"msg": "test log entry 0", "tag2": "test", ' \
+                    '"data": 5000, "some_null_field": null}'
       d.emit('message' => 'notJSON ' + json_string)
+      d.emit('log' => 'notJSON ' + json_string)
+      d.emit('msg' => 'notJSON ' + json_string)
       d.emit('message' => json_string)
-      d.emit('message' => "\t" + json_string)
-      d.emit('message' => '  ' + json_string)
+      d.emit('log' => json_string)
+      d.emit('msg' => json_string)
+      d.emit('message' => "  \r\n \t" + json_string)
+      d.emit('log' => "  \r\n \t" + json_string)
+      d.emit('msg' => "  \r\n \t" + json_string)
       d.run
     end
-    verify_log_entries(4, COMPUTE_PARAMS, '') do |entry|
-      assert entry.key?('textPayload'), 'Entry did not have textPayload'
+    log_index = 0
+    verify_log_entries(9, COMPUTE_PARAMS, '') do |entry|
+      log_index += 1
+      if log_index == 1
+        assert entry.key?('textPayload'),
+               "Entry ##{log_index} #{entry} did not have textPayload"
+      elsif log_index < 4
```

Comment:
Same comment as above. I think it's better to split the test configuration and the logic, to make adding tests in future easier

Reply:
Done.